### PR TITLE
Topic/ams 10812 implement

### PR DESCRIPTION
**Pull request by corydodt**
Implements AMS-10812: fluent-bit plugin for shipping logs to a GCS bucket.

The README describes installation, operation, and options. There will be a followup PR against ams-base-runbooks to install the plugin on hosts.

-    out_gcs.go: implement the hooks expected by fluent-bit for a plugin. Reads the options from fluent-bit config, creates objectworkers and dispatches writes to them, and cleans up on exit.
-   objectworker.go: an ObjectWorker handles the state of a connection to google cloud storage, writes there, and commits bucket objects. It also formats the filename of the GCS object using info passed from fluent-bit config options.
-  storageapi.go: interface wrapper around the google cloud storage API for Go. thin abstraction to make the API swappable (mainly for unit testing)
-    flboutputapi.go: interface wrapper for the fluent-bit plugin API for Go, similar to storageapi.go
- *_test.go: tests
- stubs_for_test.go: thin implementations of the *api.go interfaces, appropriate for unit tests.

**Note: A lot of time has passed and some changes were made, so I'm recreating the PR. Most of the new stuff is in `flbPluginFlushCtxGo`.**

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -17,14 +17,20 @@ jobs:
...
+            echo "::set-output name=release_artifact::$(make print-release-artifact)"
           else
             echo "::error::This is not a release. To create a release, tag a commit with a release version, e.g.: git tag -a -m v1.2.3 v1.2.3; git push --tags"
             false
           fi
 
+      - name: Setup Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.18.1
```
Comment (jwillhite) [inline .github/workflows/release.yaml:32]: Awesome, nice to see us using `1.18`.

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,39 @@
+module github.com/aerospike-managed-cloud-services/flb-output-gcs
+
+go 1.17
```
Comment (jwillhite) [inline go.mod:3]: For github actions we have `1.18` specified here we have `1.17`, need to update one or the other.

File: objectworker.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	"context"
+	"fmt"
+	"io"
+	"text/template"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// manages the lifetime of a gcs object
```
Comment (jwillhite) [inline objectworker.go:15]: Doc strings should always start with the name of the thing being documented, for example in this case we'd want to start with: ```// ObjectWorker ...```

ref: https://go.dev/blog/godoc

**Comment by pbaerospike** [comment]
@corydodt - Could you add some instructions to the main `README.md` on how to run the unit tests locally?  Assuming, the  reviewer test runner has no Golang experience and/or doesn't want to spend the time to read all the source code.

**Review by jwillhite — COMMENTED** [review]
This looks great, just a few comments/nits. What do you think about including at least one test that goes from a string input record to a string output record as an end-to-end test?

**Review by pbaerospike — APPROVED** [review]
Approved pending feedback on Josh's suggestions.

File: objectworker.go (hunk 3 above)
Comment (corydodt) [reply objectworker.go:15]: OK will fix as many as I find, thanks for the reference

File: go.mod (hunk 2 above)
Comment (corydodt) [reply go.mod:3]: Technically it doesn't need to match; left as-is this means go.mod itself is using the `go.mod` syntax specified by go 1.17. Per [the gomod reference](https://go.dev/doc/modules/gomod-ref) this specifies how it understands `// indirect` packages, for example, and other things that only exist in go.mod.

According to that doc there is no syntax change to `go.mod` in go 1.18.

I am nevertheless updating it just to keep it consistent.

**Pull request merged**